dt-impact-tracer/src/main/java/edu/washington/cs/dt/impact/tools/ParallelDependentTestFinder.java

```diff
@@ -0,0 +1,541 @@
+
+                // Now that we know the dependent tests in the top half, those dependencies should
+                // be handled by the knownDependencies variable.
+                List<String> orderedTests = new ArrayList<>(addOnTests);
+                orderedTests.add(dependentTestName);
+                boolean resultDifferent = isTestResultDifferent(orderedTests);
+                // Original if: if (!((!isOriginalOrder && resultDifferent) || (isOriginalOrder && !resultDifferent))) {
+                if (isOriginalOrder == resultDifferent) {
+                    dependentTestSolver(newBotList, isOriginalOrder, addOnTests, botResults.getResult(dependentTestName).result, botHalf);
+                } else {
+                    dependentTestSolver(newBotList, isOriginalOrder, new ArrayList<>(), botResults.getResult(dependentTestName).result, botHalf);
+                }
+
+                return;
+            }
+
+            // If only one half contains dependent tests, ignore all tests not in that half.
+            // If the top results match and we're looking for the before tests, then we want the top half.
+            // If the top results don't match, and we're looking for the after tests, we still want the top half.
+            if (topResultsMatch == isOriginalOrder) {
+                tests = topHalf;
+                revealed = topResults.getResult(dependentTestName).result;
+            } else {
+                tests = botHalf;
+                revealed = botResults.getResult(dependentTestName).result;
+            }
+
+            tests.removeAll(addOnTests);
+            tests.remove(tests.size() - 1);
+
+            topHalf = new LinkedList<>(tests.subList(0, tests.size() / 2));
+            botHalf = new LinkedList<>(tests.subList(tests.size() / 2, tests.size()));
+        }
+
+        if (!tests.isEmpty()) {
+            if (isOriginalOrder) {
+                addDependency(tests.get(0), revealed, revealingOrder, true);
+            } else {
+                addDependency(tests.get(0), revealed, revealingOrder, false);
+            }
+        }
+    }
+
+    /**
+     * @return A set of tests that must come AFTER the test in question.
+     */
+    private Set<String> getAfterDependencies(final String testName) {
+        return knownDependencies.entrySet().stream()
+                .filter(dependencies -> dependencies.getValue().containsAfter(testName))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toSet());
+    }
+
+    public static String dependenciesToString(final List<ParallelDependentTestFinder> dependencies) {
+        return dependencies.stream()
+                .map(ParallelDependentTestFinder::toString)
+                .reduce("", String::concat);
+    }
+
+    @Override
+    public String toString() {
+        final TestData testData = knownDependencies.get(dependentTestName);
+
+        if (testData != null) {
+            return testData.toString();
```

Comment:
Instead of returning the testData as a string, Henry's implementation should have a way to get the testData object. That way Henry can then update the precomputed dependencies' Map<String, <Set<TestData>> with the new TestData object or you could do add in this class and just return to Henry the knownDependencies (but I don't think you are currently doing that either, feel free to ignore this comment if you are).